src/lib/storage/lz4_segment/lz4_segment_iterable.hpp

```diff
@@ -42,21 +40,34 @@ class LZ4SegmentIterable : public PointAccessibleSegmentIterable<LZ4SegmentItera
   template <typename Functor>
   void _on_with_iterators(const std::shared_ptr<const PosList>& position_filter, const Functor& functor) const {
     using ValueIterator = typename std::vector<T>::const_iterator;
+    using NullValueIterator = typename pmr_vector<bool>::const_iterator;
+
+    const auto position_filter_size = position_filter->size();
 
-    auto decompressed_filtered_segment = std::vector<ValueType>(position_filter->size());
-    auto cached_block = std::vector<char>{};
-    auto cached_block_index = std::optional<size_t>{};
-    for (auto index = size_t{0u}; index < position_filter->size(); ++index) {
+    // vector storing the uncompressed values
+    auto decompressed_filtered_segment = std::vector<ValueType>(position_filter_size);
+
+    // _segment.decompress() takes the currently cached block (reference) and its id in addition to the requested
+    // element. If the requested element in not within that block, the next block will be decompressed and written to
+    // `cached_block` while the value and the new block id are returned. In case the requested element in within the
+    // cached block, the value and the input block id are returned.
+    for (auto index = size_t{0u}; index < position_filter_size; ++index) {
       const auto& position = (*position_filter)[index];
       // NOLINTNEXTLINE
       auto [value, block_index] = _segment.decompress(position.chunk_offset, cached_block_index, cached_block);
       decompressed_filtered_segment[index] = std::move(value);
       cached_block_index = block_index;
     }
 
-    auto begin = PointAccessIterator<ValueIterator>{decompressed_filtered_segment, &_segment.null_values(),
-                                                    position_filter->cbegin(), position_filter->cbegin()};
-    auto end = PointAccessIterator<ValueIterator>{decompressed_filtered_segment, &_segment.null_values(),
+    auto begin = PointAccessIterator<ValueIterator>{
+        decompressed_filtered_segment.begin(),
+        _segment.null_values() ? std::optional<NullValueIterator>{_segment.null_values()->cbegin()}
+                               : std::optional<NullValueIterator>{},
+        position_filter->cbegin(), position_filter->cbegin()};
+    auto end = PointAccessIterator<ValueIterator>{decompressed_filtered_segment.begin(),
+                                                  _segment.null_values()
+                                                      ? std::optional<NullValueIterator>{_segment.null_values()->cend()}
```

Comment:
I'm surprised that you need the std::optional here

Reply:
Narf ... I must have screwed something up. That was somewhere else.

Anyhow, does not compile if neither side of the `:` does have the `std::optional< NullValueIterator>`.

Reply:
ok